core/services/vrf/v2/listener_v2_log_listener_test.go

```diff
@@ -0,0 +1,1013 @@
+		require.NoError(t, err2)
+		th.Client.Commit()
+	}
+
+	// Blocks till now: 2 (in SetupTH) + 2 (empty blocks) + 3*5 (EmitLog + VRF req blocks) = 19
+
+	// Calling Start() after RegisterFilter() simulates a node restart after job creation, should reload Filter from db.
+	require.NoError(t, th.LogPoller.Start(th.Ctx))
+
+	// We've to replay from before VRF request log, since updateLastProcessedBlock
+	// does not internally call LogPoller.Replay
+	require.NoError(t, th.LogPoller.Replay(th.Ctx, 4))
+
+	// updateLastProcessedBlock must return the VRF req at block (7) instead of
+	// finalizedBlockNumber (16) after currLastProcessedBlock (4) passed below,
+	// as block 7 contains the earliest unfulfilled VRF request. VRF request
+	// in block 6 has been fulfilled in block 7.
+	lastProcessedBlock, err := th.Listener.updateLastProcessedBlock(th.Ctx, 4)
+	require.Nil(t, err)
+	require.Equal(t, int64(7), lastProcessedBlock)
+}
+
+/* Tests for updateLastProcessedBlock: END */
+
+/* Tests for getUnfulfilled: BEGIN
+ * TestGetUnfulfilled_NoVRFReqs
+ * TestGetUnfulfilled_NoUnfulfilledVRFReqs
+ * TestGetUnfulfilled_OneUnfulfilledVRFReq
+ * TestGetUnfulfilled_SomeUnfulfilledVRFReqs
+ * TestGetUnfulfilled_UnfulfilledNFulfilledVRFReqs
+ */
+
+func SetupGetUnfulfilledTH(t *testing.T) (*listenerV2, *utils.Big) {
+	chainID := utils.NewBig(big.NewInt(12345))
+	lggr := logger.TestLogger(t)
+	j, err := vrfcommon.ValidatedVRFSpec(testspecs.GenerateVRFSpec(testspecs.VRFSpecParams{
+		RequestedConfsDelay: 10,
+		EVMChainID:          chainID.String(),
+	}).Toml())
+	require.NoError(t, err)
+	chain := evmmocks.NewChain(t)
+
+	// Construct CoordinatorV2_X object for VRF listener
+	owner := testutils.MustNewSimTransactor(t)
+	ethDB := rawdb.NewMemoryDatabase()
+	ec := backends.NewSimulatedBackendWithDatabase(ethDB, map[common.Address]core.GenesisAccount{
+		owner.From: {
+			Balance: big.NewInt(0).Mul(big.NewInt(10), big.NewInt(1e18)),
+		},
+	}, 10e6)
+	_, _, vrfLogEmitter, err := vrf_log_emitter.DeployVRFLogEmitter(owner, ec)
+	require.NoError(t, err)
+	ec.Commit()
+	coordinatorV2, err := vrf_coordinator_v2.NewVRFCoordinatorV2(vrfLogEmitter.Address(), ec)
+	require.Nil(t, err)
+	coordinator := NewCoordinatorV2(coordinatorV2)
+
+	listener := &listenerV2{
+		respCount:   map[string]uint64{},
+		job:         j,
+		chain:       chain,
+		l:           logger.Sugared(lggr),
+		coordinator: coordinator,
+	}
+	return listener, chainID
+}
+
+func TestGetUnfulfilled_NoVRFReqs(t *testing.T) {
+	t.Parallel()
+
+	listener, chainID := SetupGetUnfulfilledTH(t)
+
+	logs := []logpoller.Log{}
+	for i := 0; i < 10; i++ {
+		logs = append(logs, logpoller.Log{
+			EvmChainId:     chainID,
+			LogIndex:       0,
+			BlockHash:      common.BigToHash(big.NewInt(int64(i))),
+			BlockNumber:    int64(i),
+			BlockTimestamp: time.Now(),
+			Topics: [][]byte{
+				[]byte("0x7299721c72d5436afdfb18da500ab7e08ac04c812d7204e1bddd322f67d8be4c"),
```

Comment:
This isn't critical, and I might be wrong on this, but `Log1` in LogEmitter doesn't have an indexed field. I _think_ `Topics` only has indexed fields (but not 100% sure). `Log2` does though. This doesn't affect the test but may be confusing for readers (like myself)

Reply:
The first item in Topics list would be the event signature right